src/state/charm_state.py

```diff
@@ -0,0 +1,211 @@
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""haproxy-operator charm state."""
+
+import itertools
+import logging
+import typing
+from enum import StrEnum
+
+# bandit flags import subprocess as a low security issue
+# We disable it here as it's simply a heads-up for potential issues
+from subprocess import STDOUT, CalledProcessError, check_output  # nosec
+
+import ops
+from charms.haproxy.v1.haproxy_route import HaproxyRouteProvider
+from charms.traefik_k8s.v1.ingress_per_unit import IngressPerUnitProvider
+from charms.traefik_k8s.v2.ingress import IngressPerAppProvider
+from pydantic import Field, ValidationError, field_validator
+from pydantic.dataclasses import dataclass
+
+from http_interface import HTTPRequirer
+
+from .exception import CharmStateValidationBaseError
+
+logger = logging.getLogger()
+
+
+class HaproxyTooManyIntegrationsError(CharmStateValidationBaseError):
+    """Exception raised when haproxy is in an invalid state with too many integrations."""
+
+
+class ProxyMode(StrEnum):
+    """StrEnum of possible http_route types.
+
+    Attrs:
+        HAPROXY_ROUTE: When haproxy-route is related.
+        INGRESS: when ingress is related.
+        INGRESS_PER_UNIT: when ingress-per-unit is related.
+        LEGACY: when reverseproxy is related.
+        NOPROXY: when haproxy should return a default page.
+        INVALID: when the charm state is invalid.
+    """
+
+    HAPROXY_ROUTE = "haproxy-route"
+    INGRESS = "ingress"
+    INGRESS_PER_UNIT = "ingress-per-unit"
+    LEGACY = "legacy"
+    NOPROXY = "noproxy"
+    INVALID = "invalid"
+
+
+class InvalidCharmConfigError(CharmStateValidationBaseError):
+    """Exception raised when a charm configuration is found to be invalid."""
+
+
+@dataclass(frozen=True)
+class CharmState:
+    """A component of charm state that contains the charm's configuration and mode.
+
+    Attributes:
+        mode: The current proxy mode of the charm.
+        global_max_connection: The maximum per-process number of concurrent connections.
+        Must be between 0 and "fs.nr_open" sysctl config.
+    """
+
+    mode: ProxyMode
+    global_max_connection: int = Field(alias="global_max_connection")
+
+    @field_validator("global_max_connection")
+    @classmethod
+    def validate_global_max_conn(cls, global_max_connection: int) -> int:
+        """Validate global_max_connection config.
+
+        Args:
+            global_max_connection: The config to validate.
+
+        Raises:
+            ValueError: When the configured value is not between 0 and "fs.file-max".
+
+        Returns:
+            int: The validated global_max_connection config.
+        """
+        if global_max_connection < 0:
```

Comment:
You can remove this and set a field constraint on the attribute definition, something like
global_max_connection: in = Field(gt=0)